pkg/loadbalancer/services.go

```diff
@@ -0,0 +1,415 @@
+	"github.com/cilium/cilium/pkg/container"
+	"github.com/cilium/cilium/pkg/source"
+	"github.com/cilium/cilium/pkg/statedb"
+	"github.com/cilium/cilium/pkg/statedb/reconciler"
+)
+
+// Services provides validated write access to the service and backend tables.
+// This is currently experimental and not in use by default.
+// See comment in [tables.go] for further context.
+type Services struct {
+	db   *statedb.DB
+	svcs statedb.RWTable[*Service]
+	bes  statedb.RWTable[*ServiceBackend]
+
+	hooks []ServiceHook
+}
+
+func NewServices(cfg Config, db *statedb.DB, svcs statedb.RWTable[*Service], bes statedb.RWTable[*ServiceBackend]) (*Services, error) {
+	if !cfg.EnableNewServices {
+		return nil, nil
+	}
+	return &Services{db, svcs, bes, nil}, nil
+}
+
+func (s *Services) IsEnabled() bool {
+	return s != nil
+}
+
+// Hook is a function that runs when a service is upserted. It can perform reads using the provided
+// ReadTxn, but for consistency it must not depend on any other data (except when the data is constant).
+// The hook can modify the service object.
+type ServiceHook func(statedb.ReadTxn, *Service)
+
+type ServiceWriteTxn struct {
+	statedb.WriteTxn
+}
+
+var (
+	ErrServiceSourceMismatch = errors.New("service exists from different source")
+	ErrServiceConflict       = errors.New("conflict with a service with same address, but different name")
+)
+
+// RegisterHook adds a new hook to modify a service when upserted. This must only be called from
+// providers/invokes and not after starting!
+func (s *Services) RegisterHook(h ServiceHook) {
+	s.hooks = append(s.hooks, h)
+}
+
+// Services returns the service table for reading.
+// Convenience method for reducing dependencies for service writers.
+func (s *Services) Services() statedb.Table[*Service] {
+	return s.svcs
+}
+
+// Services returns the backend table for reading.
+// Convenience method for reducing dependencies for service writers.
+func (s *Services) Backends() statedb.Table[*ServiceBackend] {
+	return s.bes
+}
+
+// WriteTxn returns a write transaction against services & backends and other additional
+// tables to be used with the methods of [Services]. The returned transaction MUST be
+// Abort()'ed or Commit()'ed.
+func (s *Services) WriteTxn(extraTables ...statedb.TableMeta) ServiceWriteTxn {
+	return ServiceWriteTxn{
+		s.db.WriteTxn(s.svcs, append(extraTables, s.bes)...),
+	}
+}
+
+func (s *Services) UpsertService(txn ServiceWriteTxn, params *ServiceParams) (created bool, err error) {
+	var svc Service
+
+	existing, _, found := s.svcs.First(txn, ServiceL3n4AddrIndex.Query(params.Frontend))
+	if found {
+		// Do not allow services with same address but different names to override each other.
+		if !existing.Name.Equal(params.Name) {
+			return false, ErrServiceConflict
+		}
+
+		// Do not allow overriding a service that has been created from
+		// another source. If the service is 'Unspec'
+		if existing.Source != params.Source {
+			return false, ErrServiceSourceMismatch
+		}
```

Comment:
Mostly I was trying to start out with some reasonable, but strict defaults. I have not fully thought through yet how sources that have overlapping services should interact as it's not really something we've supported before. 

Not sure why I have an unfinished comment here nor what I was thinking about with "Unspec" sources. Will clean it up.

Reply:
Thinking about this more, at least in GKE's use case, some form of prioritization, ideally configurable would be ideal. There are some cases where the same Service may transition from one data source to another, and in that case, it would be ideal if we could have a Service present in both data sources to ensure a smooth transition. Essentially that would mean if the Service came from a higher priority data source, we'd transition to that as the source of truth, but if it then disappeared from that data source, we'd transition to the next highest priority data source for that Service.

Reply:
Yeah that's kind of what we have in `pkg/source`. Unfortunately the "AllowOverwrite" is hard-coded, so it's not very useful here. Perhaps the sources should be extendable (e.g. outside Cilium OSS code if one embeds the AgentCell into a derivative project) and they should have an associated priority...

Reply:
Hmm could the priorities be different for different entities?

Reply:
> Hmm could the priorities be different for different entities?

I'm not sure what you mean by entities and what Rob has in mind, but I suppose that from the GKE perspective, it should suffice if it is possible to customise the priorities with some config file (that is, without the need to alter the code).